### ISSUE TITLE
[BUG] it took almost 1hour to :Initializing TorchBackend in DeepSpeed with backend nccl

### ISSUE THREAD

**Issue by Modas-Li** [microsoft/DeepSpeed#3318]
Help me to eliminate this time consuming ,thx

**Comment by alphanlp** [comment]
me too? do you have solve the problem?

**Comment by NicholasYoungAI** [comment]
me too

**Comment by devinzhang91** [comment]
I meet the same problem, I run in  docker container , pip list:
Package                  Version     Editable project location
------------------------ ----------- -------------------------
absl-py                  1.4.0
accelerate               0.19.0
aiohttp                  3.8.4
aiosignal                1.3.1
antlr4-python3-runtime   4.9.3
appdirs                  1.4.4
async-timeout            4.0.2
attrs                    23.1.0
blinker                  1.6.2
certifi                  2023.5.7
chardet                  5.1.0
charset-normalizer       3.1.0
click                    8.1.3
cmake                    3.26.3
colorama                 0.4.6
cpm-kernels              1.0.11
DataProperty             0.55.1
datasets                 2.10.1
deepspeed                0.8.3
dill                     0.3.4
docker-pycreds           0.4.0
einops                   0.6.1
evaluate                 0.4.0
filelock                 3.12.0
flash-attn               1.0.4
Flask                    2.3.2
Flask-Cors               3.0.10
frozenlist               1.3.3
fsspec                   2023.5.0
gitdb                    4.0.10
GitPython                3.1.31
hjson                    3.1.0
huggingface-hub          0.14.1
icetk                    0.0.7
idna                     3.4
importlib-metadata       6.6.0
itsdangerous             2.1.2
Jinja2                   3.1.2
joblib                   1.2.0
jsonlines                3.1.0
lit                      16.0.3
lm-eval                  0.3.0
lmflow                   0.0.1       /root/LMFlow/src
MarkupSafe               2.1.2
mbstrdecoder             1.1.2
mpi4py                   3.1.4
mpmath                   1.3.0
multidict                6.0.4
multiprocess             0.70.12.2
networkx                 3.1
ninja                    1.11.1
nltk                     3.8.1
numexpr                  2.8.4
numpy                    1.24.2
nvidia-cublas-cu11       11.10.3.66
nvidia-cuda-cupti-cu11   11.7.101
nvidia-cuda-nvrtc-cu11   11.7.99
nvidia-cuda-runtime-cu11 11.7.99
nvidia-cudnn-cu11        8.5.0.96
nvidia-cufft-cu11        10.9.0.58
nvidia-curand-cu11       10.2.10.91
nvidia-cusolver-cu11     11.4.0.1
nvidia-cusparse-cu11     11.7.4.91
nvidia-nccl-cu11         2.14.3
nvidia-nvtx-cu11         11.7.91
omegaconf                2.3.0
openai                   0.27.6
packaging                23.1
pandas                   2.0.1
pathtools                0.1.2
pathvalidate             2.5.2
peft                     0.3.0.dev0
Pillow                   9.5.0
pip                      23.0.1
portalocker              2.7.0
protobuf                 3.18.3
psutil                   5.9.5
py-cpuinfo               9.0.0
pyarrow                  12.0.0
pybind11                 2.10.4
pycountry                22.3.5
pydantic                 1.10.7
pytablewriter            0.64.2
python-dateutil          2.8.2
pytz                     2023.3
PyYAML                   6.0
regex                    2023.5.5
requests                 2.30.0
responses                0.18.0
rouge-score              0.1.2
sacrebleu                1.5.0
scikit-learn             1.2.2
scipy                    1.10.1
sentencepiece            0.1.99
sentry-sdk               1.22.2
setproctitle             1.3.2
setuptools               66.0.0
six                      1.16.0
smmap                    5.0.0
sqlitedict               2.1.0
sympy                    1.12
tabledata                1.3.1
tcolorpy                 0.1.3
threadpoolctl            3.1.0
tokenizers               0.13.3
torch                    2.0.0
torchvision              0.15.1
tqdm                     4.65.0
tqdm-multiprocess        0.0.11
transformers             4.28.0.dev0
triton                   2.0.0
trl                      0.4.2.dev0
typepy                   1.3.0
typing_extensions        4.5.0
tzdata                   2023.3
urllib3                  1.26.15
wandb                    0.14.0
Werkzeug                 2.3.4
wheel                    0.38.4
xxhash                   3.2.0
yarl                     1.9.2
zipp                     3.15.0
zstandard                0.21.0

**Comment by devinzhang91** [comment]
> me too

Is your machine connected to the internet? I also got stuck in an offline environment at the beginning, but after the machines connected to the internet,  quickly passed.

**Comment by cxxz** [comment]
The server I'm using is unable to connect to the internet. Can I configure DeepSpeed to prevent the offline environment from slowing down the NCCL initialization process?

**Comment by devinzhang91** [comment]
> The server I'm using is unable to connect to the internet. Can I configure DeepSpeed to prevent the offline environment from slowing down the NCCL initialization process?

I  think ``` deepspeed.init_distributed ``` this API cause this pormble, find it and try to set timeout  as 10 sec.
https://deepspeed.readthedocs.io/en/latest/initialize.html#deepspeed.init_distributed

**Comment by askxiaozhang** [comment]
> Hi @Modas-Li, @alphanlp, and @NicholasYoungAI can you provide more information about your system and set up? Thanks

I have same problem when I t started DeepSpeed-Chat that bash training/step2_reward_model_finetuning/training_scripts\single_node\run.sh 

this is detail:  (it waste my exceed about 6hours)can't go on
[2023-07-13 14:51:33,751] [WARNING] [runner.py:191:fetch_hostfile] Unable to find hostfile, will proceed with training with local resources only.
Detected CUDA_VISIBLE_DEVICES=1,2,3: setting --include=localhost:1,2,3
[2023-07-13 14:51:33,954] [INFO] [runner.py:541:main] cmd = /home/zhangchang/.conda/envs/chatdeep/bin/python -u -m deepspeed.launcher.launch --world_info=eyJsb2NhbGhvc3QiOiBbMSwgMiwgM119 --master_addr=127.0.0.1 --master_port=29500 --enable_each_rank_log=None main.py --data_path local/rm-static --model_name_or_path THUDM/chatglm-6b --num_padding_at_beginning 1 --per_device_train_batch_size 4 --per_device_eval_batch_size 4 --max_seq_len 512 --learning_rate 5e-5 --weight_decay 0.1 --num_train_epochs 1 --disable_dropout --gradient_accumulation_steps 1 --lr_scheduler_type cosine --num_warmup_steps 0 --seed 1234 --zero_stage 3 --deepspeed --output_dir ./output2
[2023-07-13 14:51:35,266] [INFO] [launch.py:222:main] 0 NCCL_IBEXT_DISABLE=1
[2023-07-13 14:51:35,267] [INFO] [launch.py:222:main] 0 NCCL_IB_DISABLE=1
[2023-07-13 14:51:35,267] [INFO] [launch.py:229:main] WORLD INFO DICT: {'localhost': [1, 2, 3]}
[2023-07-13 14:51:35,267] [INFO] [launch.py:235:main] nnodes=1, num_local_procs=3, node_rank=0
[2023-07-13 14:51:35,267] [INFO] [launch.py:246:main] global_rank_mapping=defaultdict(<class 'list'>, {'localhost': [0, 1, 2]})
[2023-07-13 14:51:35,267] [INFO] [launch.py:247:main] dist_world_size=3
[2023-07-13 14:51:35,267] [INFO] [launch.py:249:main] Setting CUDA_VISIBLE_DEVICES=1,2,3
[2023-07-13 14:51:37,178] [INFO] [comm.py:622:init_distributed] Initializing TorchBackend in DeepSpeed with backend nccl

**Comment by clumsy** [comment]
Can you please try collecting `top -n` results using `py-spy` while this issue is happening? This should help understanding what the process executing during this time. It would also be nice if you can generate a flame graph too.

**Comment by molly-smith** [comment]
Hi all, NCCL and internet are required to run DeepSpeed. We do not support running without these. Closing this issue. Thanks.
https://github.com/microsoft/DeepSpeed/issues/4104

**Comment by kyriemao** [comment]
This problem occur when I using transformers 4.25, but disappear when I upgrade the transformers library to 4.26

**Comment by YeFD** [comment]
When setting `--main_process_port 0`, I encountered the same issue. Removing it or specifying a port other than 0 allowed it to run normally. 
I have no idea to verify if there's a bug for automatic port selection. If anyone has solved this problem, please let me know. My deepspeed version is 0.16.4.

**Comment by VincentYCYao** [comment]
> When setting `--main_process_port 0`, I encountered the same issue. Removing it or specifying a port other than 0 allowed it to run normally. I have no idea to verify if there's a bug for automatic port selection. If anyone has solved this problem, please let me know. My deepspeed version is 0.16.4.

Exactly the same issue and solution with DeepSpeed 0.16.5.

Use `--main_process_port <some port>` instead of `--main_process_port 0`
